plugins/lookup/reverse_lookup.py

```diff
@@ -0,0 +1,175 @@
+description:
+    - Look up hostnames for IP addresses using DNS reverse lookup.
+options:
+    _terms:
+        description:
+            - IP address(es) to look up.
+        type: list
+        elements: str
+        required: true
+    query_retry:
+        description:
+            - Number of retries for DNS query timeouts.
+        type: int
+        default: 3
+    query_timeout:
+        description:
+            - Timeout per DNS query in seconds.
+        type: float
+        default: 10
+    server:
+        description:
+            - The DNS server(s) to use to look up the result. Must be a list of one or more IP addresses.
+            - By default, the system's standard resolver is used.
+        type: list
+        elements: str
+    servfail_retries:
+        description:
+            - How often to retry on SERVFAIL errors.
+        type: int
+        default: 0
+notes:
+    - Note that when using this lookup plugin with V(lookup(\)), and the result is a one-element list,
+      Ansible simply returns the one element not as a list. Since this behavior is surprising and
+      can cause problems, it is better to use V(query(\)) instead of V(lookup(\)). See the examples
+      and also R(Forcing lookups to return lists, query) in the Ansible documentation.
+'''
+
+EXAMPLES = """
+- name: Look up hostname of IPv4 address
+  ansible.builtin.debug:
+    msg: "{{ query('community.dns.reverse_lookup', '192.168.1.1') }}"
+
+- name: Look up hostname of IPv6 address
+  ansible.builtin.debug:
+    msg: "{{ query('community.dns.reverse_lookup', '1:2:3::4') }}"
+"""
+
+RETURN = """
+_result:
+    description:
+        - The hostname(s) returned for the queried IP addresses.
+        - If multiple IP addresses are queried in O(_terms), the resulting lists have been concatenated.
+    type: list
+    elements: str
+    sample:
+        - example.com
+        - example.org
+"""
+
+from ansible.errors import AnsibleLookupError
+from ansible.plugins.lookup import LookupBase
+from ansible.module_utils.common.text.converters import to_text
+
+from ansible_collections.community.dns.plugins.module_utils.ips import (
+    is_ip_address,
+)
+
+from ansible_collections.community.dns.plugins.module_utils.resolver import (
+    SimpleResolver,
+)
+
+from ansible_collections.community.dns.plugins.plugin_utils.ips import (
+    assert_requirements_present as assert_requirements_present_ipaddress,
+)
+
+from ansible_collections.community.dns.plugins.plugin_utils.resolver import (
+    assert_requirements_present as assert_requirements_present_dnspython,
+    guarded_run,
+)
+
+try:
+    import dns.resolver
+except ImportError:
+    # handled by assert_requirements_present_dnspython
+    pass
+
+try:
+    import ipaddress
+except ImportError:
+    # handled by assert_requirements_present
+    pass
+
+
+class LookupModule(LookupBase):
+    @staticmethod
+    def _resolve(resolver, name, rdtype, server_addresses):
```

Comment:
```suggestion
    def _resolve(resolver, name: str, rdtype: int, server_addresses: Optional[List[str]]): -> List[str]:
```

Reply:
I think the type hint for the `resolver` parameter would be `SimpleResolver` right? so `resolver: SimpleResolver`